### ISSUE TITLE
Source or derive a historical view of county-level data

### ISSUE THREAD

**Issue by lisac-usds** [nickrobison-usds/demand-modeler#91]
## Background

The [COVID Tracking Project](https://covidtracking.com/data/) links to county-level data provided by states.  Data points include: positive cases, negative cases, pending tests, hospitalizations, and deaths. 

States vary in which of these data points they report, and the format of their report. The formats range from structured (APIs and csv files) to unstructured (pdf reports, news articles, images of the figures as displayed on a webpage). In most cases, the reports are a snapshot for a single date.

To see the county-level data for a state, follow the "Best current data source for [state]" links under each state's summary table on [COVID Tracking Project](https://covidtracking.com/data/) . For example, here are the links for [Alaska](http://dhss.alaska.gov/dph/Epi/id/Pages/COVID-19/monitoring.aspx), [California](https://www.latimes.com/projects/california-coronavirus-cases-tracking-outbreak/), and [Connecticut](https://portal.ct.gov/Coronavirus) (find the "Latest COVID-19 Testing Data in Connecticut" pdf on the resulting page).

## Objective
Train predictive models using a *historical view* of this data.

## Task
Source or derive a single dataset that reliably displays a *historical view* across the counties of the states. 

---
### Sourcing a dataset

It's possible what we're looking for already exists, and we don't know where to look.
What is already out there, and how can we verify their accuracy? 

We are currently pulling from our data from a CSBS data source that they update hourly. Their daily historical data is stored on [GitHub](https://github.com/CSBSData/COVID-19). However, in generating a daily report from their data we have noticed that it is often 48+ hours out of date from what the states are reporting.

---
### Deriving a single dataset

A number of approaches to this.

#### Option A

Considering the variation in how states are reporting their county-level data, develop a historical dataset per state, with potentially unique implementations for each state. Across these implementations, standardize on a csv format and naming schema.

Prioritize the states with relatively high confirmed cases (potentially WA, CA, and NY), also considering those which are friendliest for scraping (eg CO seems to offer an API).

Could we divide-and-conquer by having people sign up for a specific state?

#### Other Options

Open to recommendations. 

## Where we are at

Currently we are downloading the current days csv manually adjusting the top 10 states (ordered by confirmed count) and loading that into our data base. Kelvin and Tim are working on a more scalable way to ingest the data for scrappers to be setup

##
CSV format:
```
countyName = row[0]
stateName = row[1]
confirmed = row[2]
lastUpdate = row[8]
STATEFP = row[9]
COUNTYFP = row[10]
```

**Comment by timothybest-usds** [comment]
Example NY top 10 counties 3-22:

|county|CSBC|ny.gov|
| ------------- |:-------------:| -----:|
|New York|8115|9045| 
|Westchester|1387|1873| 
|Nassau|1234|1900| 
|Suffolk|552|1034| 
|Rockland|262|455| 
|Orange|163|247| 
|Albany|88|123| 
|Nassau|1234|82| 
|Monroe|46|57| 
|Erie|56|56| 
|Saratoga|35|41|

**Comment by lisac-usds** [comment]
to investigate:  

1. The University of Virginia's COVID-19 Surveillance Dashboard https://nssac.bii.virginia.edu/covid-19/dashboard/ .  This could fit into the "Sourcing a dataset" path. 

2. somebody already put this data into a csv?  https://docs.google.com/spreadsheets/d/1jLAjzYMcsPo71qu5uBxjT_dAzj8vxZfvnTMR-Xp3-v8/edit#gid=1076558078

3. https://github.com/mimisun/covid19 

Thanks to Boris for tracking these down.

**Comment by lisac-usds** [comment]
The University of Virginia's dashboard seems really useful. I need some to think through next steps, will outline what I'm thinking before implementing anything. In the meantime, here's what I found:

A sample of data dumps from the University of Virginia dashboard, one row per state. County figures, if available, are displayed as a string in the "Last Update" column. 

Columns: Place (the state), Region (United States), Confirmed Cases, Deaths, Recovered, Active, Last Update (may include description of the county figures).

I think the county figures reflect the confirmed cases, not the active cases. Looking at Indiana for the 2020-03-21 snapshot: 
```
"Indiana","United States","129","4","0","125","2020-03-22 03:00:00 * CTY: Marion 46; Hamilton 10; Johnson 7; Lake 7; St. Joseph 6; Hendricks 6; Howard 5; Floyd 5; Clark 4; Allen 4; Delaware  2; Vigo 2; Franklin 2; LaPorte 2; Boone 2; Tippecanoe 2; Noble 1; Vanderburgh 1; Owen 1; Wayne 1; Grant 1; Greene 1; Jennings 1; Hancock 1; Scott 1; Madison 1; Fayette 1; Bartholomew 1; Elkhart 1; Adams 1; Shelby 1; Monroe 1; Wells 1"
```
```
>>> description = "Marion 46; Hamilton 10; Johnson 7; Lake 7; St. Joseph 6; Hendricks 6; Howard 5; Floyd 5; Clark 4; Allen 4; Delaware  2; Vigo 2; Franklin 2; LaPorte 2; Boone 2; Tippecanoe 2; Noble 1; Vanderburgh 1; Owen 1; Wayne 1; Grant 1; Greene 1; Jennings 1; Hancock 1; Scott 1; Madison 1; Fayette 1; Bartholomew 1; Elkhart 1; Adams 1; Shelby 1; Monroe 1; Wells 1"
>>> counties = description.split(';')
>>> total = 0
>>> for c in counties:
...     total += int(c.strip().split()[-1].strip())
... 
>>> total
129
```

Samples of the data as pulled from the UVA dashboard:
(Uploaded with .txt extension only b/c .csv isn't a supported file type for uploads on a github issue.)
[uva-2020-03-04.csv.txt](https://github.com/nickrobison-usds/demand-modeler/files/4366847/uva-2020-03-04.csv.txt)
[uva-2020-03-05.csv.txt](https://github.com/nickrobison-usds/demand-modeler/files/4366846/uva-2020-03-05.csv.txt)
[uva-2020-03-06.csv.txt](https://github.com/nickrobison-usds/demand-modeler/files/4366845/uva-2020-03-06.csv.txt)
[uva-2020-03-07.csv.txt](https://github.com/nickrobison-usds/demand-modeler/files/4366843/uva-2020-03-07.csv.txt)
[...]
[uva-2020-03-21.csv.txt](https://github.com/nickrobison-usds/demand-modeler/files/4366848/uva-2020-03-21.csv.txt)

cc @timothybest-usds 🕷

**Comment by lisac-usds** [comment]
I'm going to do this little exercise in python, just to explore. Hoping to check in in an hour, but not confident. 

Task: Use the UVA dashboard data to create a csv that shows for New York the count of confirmed cases, by county, by date. Use timeframe 2020-03-04 to 2020-03-07.

output:  a csv structured like this:
|county name|2020-03-04|2020-03-05|2020-03-06|2020-03-07|
|-|-|-|-|-|
|Allentown|0|0|0|1|
|Beantown|1|1|3|4|
|Corning|1|4|4|5
|Fairmont|0|0|2|4|
|total|2|5|9|14|

**Comment by timothybest-usds** [comment]
Still working on formalizing how to ingest the data into our backend. Right now I think the easiest path forward is to copy our existing data ingestion pattern (CSVs in [data/](https://github.com/nickrobison-usds/demand-modeler/tree/master/data)).

Currently the format is `County Name,State Name,Confirmed,New,Death,Fatality Rate,Latitude,Longitude,Last Update,STATEFP,COUNTYFP,Population Estimate 2018 (Ths.),Infection Rate` where we only care about:

```
countyName = row[0]
stateName = row[1]
confirmed = row[2]
lastUpdate = row[8]
```

When we load these data files into our application we associate the county [FIPSID](https://en.wikipedia.org/wiki/Federal_Information_Processing_Standard_state_code) with the `countyName` and `stateName` in the CSV and have a unique constraint that only allows one name to be associated with one county

**Comment by lisac-usds** [comment]
> Long term it will be advantages to collect more information (both at county and state level) other then case confirmations and deaths. Please edit this issue with what column that data will be in

From what I can tell, the UVA dataset displays the confirmation counts per county, but not other figures per count (eg new, recovered, deaths. Going to proceed with trying to produce output for the counties of New York state from March 3-7. Best case: from there, include counties of 2-3 other states; and if that goes well, extend the timeframe (would involve getting more data dumps from the UVA dashboard, which is doable).

**Comment by lisac-usds** [comment]
update: I have a script for the exercise I'd described above. Don't know whether it's going to fit in nicely with current work, could use feedback, but not something I'm looking for tonight. 
Sample derived file for Mar3-8 for only New York state in https://github.com/nickrobison-usds/demand-modeler/pull/96#issuecomment-602380177.   

Going to do some cleanups for my script and then quit for the night, will check back tomorrow evening for any tasks I can help with.

**Comment by lisac-usds** [comment]
Closed #96, as it didn't seem useful. 

Other things to consider:
- https://github.com/CSSEGISandData/COVID-19 (daily reports and time series from)
- https://coronadatascraper.com/#features.json
